### DISCUSSION TITLE
DS4windows implementation

### DISCUSSION THREAD

**Discussion by Phantom2332** [ClassicOldSong/Apollo#58] [General]
So recently I started using DS4windows for better tuning a BSP controller that I have for using Apollo/Artemis, however with each boot up Apollo creates a virtual controller with different PID/MAC that makes DS4windows not detecting it properly or having difficulties with it, apparently having a fixed PID of the virtual controller created by sunshine(Apollo) would make this handling easier.

More info: https://github.com/schmaldeo/DS4Windows/issues/24#issuecomment-2381384828

**Comment by ClassicOldSong** [comment]
If you're not using the gyro function in your DS4 or compatible controllers, you can change the emulated controller to Xbox only in the input settings.

**Comment by Phantom2332** [comment]
I know that option exist but since controller settings are nonexistent with Sunshine it would be nice to be able to use a software like that to configure it,for example the controller that I use,has embedded dead zones even at 0% on it and to make it work properly the only way is to use DS4windows.

Implementing an unique PID for the virtual controller created by sunshine would make the handling of ds4windows easier, since each time it boots it creates a new one with a random PID

**Comment by ClassicOldSong** [comment]
Just checked the code, unfortunately the serial number cannot be assigned manually for [ViGEmBus](https://github.com/nefarius/ViGEmBus) which Sunshine uses for gamepad emulation...

And as HidHide is developed by the same guy who made ViGEmBus, there're actually lots of conflicts using the emulated controllers with HidHide outside of their desired use cases, from my own experience with HandHeldCompanion.

The PID assigning you mentioned cannot be done either, as VID/PID is the mandatory way to determin the type of the connected device. Write a randomly generated PID could cause lots of unwanted side effects.

Since ViGEmBus has stopped maintanance, I actually wanted to solve this issue with another method, which should work nice with HidHide and DS4Windows, but I don't have that much time to work on everything, so it might not be feasiable in the near future.

**Reply by Phantom2332** [reply]
No problem,just wanted you to know in case something easy was possible, looking forward to see it as a improvement when you consider it opportune 👍 keep up the good work this already feels like 1.0 sunshine

**Reply by ClassicOldSong** [reply]
And... Unfortunately, DS4Windows also uses ViGEmBus that is making things even more complicated...

**Reply by Phantom2332** [reply]
Yh and he dropped maintenance,the link that I passed you is of the new manteiner,he already implemented some kind of support,but because of the code structure that you mentioned is a difficult issue,that's why I proposed him to maybe get in touch with you in case it would make things easier

**Reply by Phantom2332** [reply]
Btw thinking about it mainly I'm missing the ability to set an anti-deadzone for games that doesn't allow to modify it properly,would be possible to set a negative deadzone in Artemis?

**Reply by ClassicOldSong** [reply]
You're in luck, I just made seekbar able to accept negative values lol

**Reply by Phantom2332** [reply]
Let's go! Ty, waiting for the release 🤣👍

**Reply by ClassicOldSong** [reply]
Released: https://github.com/ClassicOldSong/moonlight-android/releases/tag/v12.1-artemis-241009

**Reply by dertaika** [reply]
bruh, ran into the same detection issue with DS4Windows and the Sunshine/Apollo virtual controller — random PID on every boot means DS4Windows either grabs the wrong device or loses track of it after a restart. Switching the emulated controller to Xbox-only in [https://ds4-windows.io/](https://ds4-windows.io/)  input settings works around it if gyro isn't needed, but that's not an option for controllers where the deadzone/curve tuning only lives in DS4Windows. The anti-deadzone addition to Artemis is a solid interim fix for that exact case — appreciate the quick turnaround on it.

Sorry for bump in old thread tho
